Source/Core/VideoCommon/AsyncRequests.h

```diff
@@ -0,0 +1,90 @@
+// Copyright 2013 Dolphin Emulator Project
+// Licensed under GPLv2
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <atomic>
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+
+#include "Common/CommonTypes.h"
+
+class AsyncRequests
+{
+public:
+	struct Event
+	{
+		enum Type
+		{
+			EFB_POKE_COLOR,
+			EFB_POKE_Z,
+			EFB_PEEK_COLOR,
+			EFB_PEEK_Z,
+			SWAP_EVENT,
+			BBOX_READ,
+			PERF_QUERY,
+		} type;
+		u64 time;
+
+		union
+		{
+			struct
+			{
+				u16 x;
+				u16 y;
+				u32 data;
+			} efb_poke;
+
+			struct
+			{
+				u16 x;
+				u16 y;
+				u32* data;
+			} efb_peek;
+
+			struct
+			{
+				u32 xfbAddr;
+				u32 fbWidth;
+				u32 fbStride;
+				u32 fbHeight;
+			} swap_event;
+
+			struct
+			{
+				int index;
+				u16* data;
+			} bbox;
+
+			struct
+			{
+			} perf_query;
+		};
+	};
+
+	AsyncRequests();
+
+	void PullEvents() { if (!m_empty) PullEventsInternal(); }
+	void PushEvent(const Event& event, bool blocking = false);
+	void SetEnable(bool enable);
+	void SetPassthrough(bool enable);
+
+	static AsyncRequests* GetInstance() { return &s_singleton; }
+
+private:
+	void PullEventsInternal();
+	void HandleEvent(const Event& e);
+
+	static AsyncRequests s_singleton;
+
+	std::atomic<bool> m_empty;
```

Comment:
Please use .store()/.load() explicity on atomic values to make it clear that they're atomics at usage time.